osquery/tables/system/windows/programs.cpp

```diff
@@ -25,35 +25,47 @@ void keyEnumPrograms(const std::string& key,
   QueryData regResults;
   queryKey(key, regResults);
   for (const auto& rKey : regResults) {
+    // Each subkey represents a program, skip if not a subkey
     if (rKey.at("type") != "subkey") {
       continue;
     }
+
+    // Ensure we only process each program one time
+    const auto& fullProgramName = rKey.at("path");
+    if (processed.find(fullProgramName) != processed.end()) {
+      continue;
+    }
+    processed.insert(fullProgramName);
+
+    // Query additional information about the program
     QueryData appResults;
-    const auto& subkey = rKey.at("path");
-    // make sure it's a sane uninstall key
+    queryKey(fullProgramName, appResults);
+    Row r;
+
+    // Attempt to derive the program identifying GUID
+    std::string identifyingNumber("");
     boost::smatch matches;
     boost::regex expression(
         "({[a-fA-F0-9]+-[a-fA-F0-9]+-[a-fA-F0-9]+-[a-fA-F0-9]+-[a-fA-F0-9]+})"
         "$");
-    if (!boost::regex_search(subkey, matches, expression)) {
-      continue;
+    if (boost::regex_search(fullProgramName, matches, expression)) {
+      identifyingNumber = matches[0];
+      r["identifying_number"] = identifyingNumber;
     }
-    // Ensure we only process a program once
-    auto processGuid = matches[0];
-    if (processed.find(processGuid) != processed.end()) {
-      continue;
-    }
-    processed.insert(processGuid);
-    queryKey(subkey, appResults);
-    Row r;
-    r["identifying_number"] = processGuid;
+
     for (const auto& aKey : appResults) {
+      if (identifyingNumber.empty() && aKey.at("name") == "BundleIdentifier") {
```

Comment:
This series of checks isn't very efficient. 
```cpp
auto name = aKey.find("name");
if (name->second == "ThingYouWantToCompare") {
...
```
Then you're not traversing the RB tree for each if statement.